components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/jwt/JWKSBasedJWTValidator.java

```diff
@@ -156,13 +163,35 @@ private void checkCertificateValidity(String jwksUri, SignedJWT jwt) throws Malf
             }
 
             if (x509Certificate != null) {
-                x509Certificate.checkValidity();
+
+                String alg = jwt.getHeader().getAlgorithm().getName();
+                if (log.isDebugEnabled()) {
+                    log.debug("Signature Algorithm found in the JWT Header: " + alg);
+                }
+                if (alg.indexOf(RS) == 0 || alg.indexOf(PS) == 0) {
+                    // At this point 'x509Certificate' will never be null.
+                    PublicKey publicKey = x509Certificate.getPublicKey();
+                    if (publicKey instanceof RSAPublicKey) {
+                        x509Certificate.checkValidity();
+                    } else {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Public key is not an RSA public key.");
+                        }
+                        return false;
+                    }
```

Comment:
Shall we get this into a private method for more clarity?

Reply:
Fixed with: https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2701/commits/ec2a29944b5e4d3048848d8eeee4afa9dedc0cbc